### PR TITLE
Explainers

### PR DESCRIPTION
**Pull request by dufourc1**

### CODE REVIEW

File: torch_geometric/nn/models/explainer_algorithms.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+        Returns true if the explainer supports the settings.
+
+        Args:
+            *args: the user-defined explanation settings.
+            **kwargs: the user-defined explanation settings.
+        """
+        pass
+
+
```
Comment (RexYing) [inline torch_geometric/nn/models/explainer_algorithms.py]: Maybe we could try the shapley value approach (captum has it; or we could potentially use our own), since it's a common ask from practioners?

If it requires too much changes, then don't worry. We'll have a separate issue for this feature then.

File: torch_geometric/nn/models/explainer_algorithms.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        This method compute the loss to be used for the explanation algorithm.
+        Subclasses should override this method to define their own loss.
+
+        Args:
+            inputs (Tuple[torch.Tensor, ...]): the inputs of the GNN.
+            output (torch.Tensor): the output of the GNN.
+            target (torch.Tensor): the target of the GNN.
+        """
+
+    def _set_loss(
```
Comment (RexYing) [inline torch_geometric/nn/models/explainer_algorithms.py]: maybe `set_objective` would better describe it? `set_loss` might be thought of as related to the GNN loss.

File: torch_geometric/nn/models/explainer_algorithms.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    # TODO: how to work with the inputs for all kind of models?
+    # (index, features...)
+    @abstractmethod
+    def explain(
+        self,
+        inputs: Tuple[torch.Tensor, ...],
+        target: torch.Tensor,
+        model: torch.nn.Module,
+    ) -> Tuple[torch.Tensor, ...]:
+        r"""This method compute the explanation of the GNN and return
```
Comment (RexYing) [inline torch_geometric/nn/models/explainer_algorithms.py]:
```suggestion
        r"""This method computes the explanation of the GNN and return
```

File: torch_geometric/nn/models/explainer_algorithms.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        Args:
+            inputs (Tuple[torch.Tensor, ...]): the inputs of the GNN.
+            target (torch.Tensor): the target of the GNN.
+            model (torch.nn.Module): the GNN to explain.
+        """
+        pass
+
+    # TODO: improve on just passing a general dict
+    @abstractmethod
+    def support(self, *args, **kwargs) -> bool:
```
Comment (RexYing) [inline torch_geometric/nn/models/explainer_algorithms.py]:
```suggestion
    def supports(self, *args, **kwargs) -> bool:
```

File: torch_geometric/nn/models/explainer_settings.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
+from typing import Tuple
+
+import torch
+
+from torch_geometric.nn.models.explainer_algorithms import ExplainerAlgorithm
+
+
+class ExplainerSetter(torch.nn.Module):
+    r"""A user configuration class for the instance level-explanation
```
Comment (RexYing) [inline torch_geometric/nn/models/explainer_settings.py]:
```suggestion
    r"""A user configuration class for the instance-level explanation
```

File: torch_geometric/nn/models/explainer_settings.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    of Graph Neural Networks, *e.g.* :class:`~torch_geometric.nn.GCN`.
+
+    Args:
+        explanation_type (str): the type of explanation to be computed.
```
Comment (RexYing) [inline torch_geometric/nn/models/explainer_settings.py]: this might be related to objectives. e.g. for `model`, it makes sense to use MI as objective; for `phenomenon`, we might use the logit corresponding to the ground truth class.
So when we construct say `Saliency(ExplainerAlgorithm)`, we can take in this and create the objective accordingly.

**Review by RexYing — COMMENTED** [review]
thanks for the first draft!
